pkgs/fc/ceph/src/fc/ceph/util/opslog.py

```diff
@@ -0,0 +1,130 @@
+            f.write(
+                json.dumps(
+                    dict(
+                        last_gced_day=self.last_gced_day.strftime("%Y-%m-%d"),
+                        last_processed_datetime=self.last_processed_datetime.strftime(
+                            "%Y-%m-%dT%H"
+                        ),
+                    )
+                )
+            )
+
+
+class OpsLog:
+    def __init__(self, state_file: Path):
+        self.state_file = state_file
+        self.opslog_ptrn = re.compile(
+            r"^[\d]{4}-[\d]{2}-[\d]{2}-[\d]{2}-[A-Za-z0-9.-]+$"
+        )
+
+        if not self.state_file.exists():
+            self.opslog_state = OpsLogState(
+                datetime.today() - timedelta(days=1),
+                date.today() - timedelta(days=2),
+            )
+            self.opslog_state.write_to(self.state_file)
+        else:
+            self.opslog_state = OpsLogState.read_from(self.state_file)
+
+    @contextmanager
+    def get_pending_stats_by_day(self):
+        """
+        Provide pending log objects from the timespan between the last processed hour and
+        now. The current hour is left out because this is only running once a day once all stats
+        were collected.
+
+        It's crucial to commit the processed data within the context manager: if an
+        exception is thrown, the processed days won't be saved for a retry.
+        """
+
+        # Logs are recorded by hour
+        max_date = datetime.now().replace(microsecond=0, second=0, minute=0)
+        start = self.opslog_state.last_processed_datetime + timedelta(hours=1)
+
+        stats_by_day = {}
+
+        start_day = start.date()
+        end_day = max_date.date()
+
+        day = start_day
+        while day <= end_day:
+            logs = sorted(
+                entry
+                for entry in run.json.radosgw_admin(
+                    "log", "list", f"--date={day.strftime('%Y-%m-%d')}"
+                )
+                if self.opslog_ptrn.match(entry)
+            )
+
+            # The format is YYYY-mm-dd-HH-<uuid>-<bucket>. Don't process hours
+            # of a day that were already processed.
+            log_hours = [int(x[11:13]) for x in logs]
+
+            if day == end_day:
+                logs = logs[: log_hours.index(max_date.hour)]
+            if day == start_day:
+                logs = logs[log_hours.index(start.hour) :]
+
+            if logs:
+                try:
+                    stats_by_day[day].extend(logs)
+                except KeyError:
+                    stats_by_day[day] = logs
+
+            day += timedelta(days=1)
+
+        self.opslog_state.last_processed_datetime = max_date - timedelta(
+            hours=1
+        )
+
+        yield stats_by_day
+
+        self.opslog_state.write_to(self.state_file)
+
+    def gc_log_objects(self):
+        day = self.opslog_state.last_gced_day + timedelta(days=1)
+        end = self.opslog_state.last_processed_datetime.date() - timedelta(
```

Comment:
lets have a small guard that ensures that end is always at most yesterday and never today or in the future.